code-review.js

```diff
@@ -1,60 +1,111 @@
  * Implemented as a hash of arrays where the hash keys are priority values.
+ *
+ * @class
  */
-function PriorityQueue(size) {
+function PriorityQueue() {
 	this.store = {};	// keys are priorities, values are arrays of elements
 	this.count = 0;
 
-	// adds an item
-	// priority must be an integer (higher value has higher priority)
+
+	/**
+	 * Adds an item
+	 * priority must be a positive integer (higher value has higher priority)
+	 * 
+	 * @param {*} value - The Item we wish to change
+	 * @param {Number} priority - Positive integer 
+	 */
 	this.add = function(value, priority) {
-		if (this.store[priority] == undefined)
+		if (this.store[priority] == undefined) {
 			this.store[priority] = [];
+		}
 
 		this.store[priority].push(value);
 		this.count++;
 	};
 
-	// returns the oldest-added value with the highest priority
-	this.Pop = function() {
-		maxKey = Math.max(Object.keys(this.store));
+	/**
+	 * Returns the oldest-added value with the highest priority
+	 */
+	this.pop = function() {
+		if (this.count === 0) {
+			return false;
+		}
+		var maxKey = Math.max.apply(null, this.getAllPriorities());
+		var poppedValue = this.store[maxKey].shift();
 		this.count--;
-		return this.store[maxKey].shift();
+		this.removeEmptyPriority(maxKey);
+
+		return poppedValue
 	};
 
+	/**
+	 * Returns the number of elements in queue
+	 * @returns {Number} 
+	 */
 	this.length = function() {
 		return this.count;
-	}
-}
+	};
+};
 
-PriorityQueue.prototype.get_all_priorities = function() {
+/**
+ * Returns an array of numerical priorities
+ * @returns {Array}
+ */
+PriorityQueue.prototype.getAllPriorities = function() {
 	return Object.keys(this.store);
-}
+};
 
-// iterates through all the queue elements in priority-then-FIFO order
-PriorityQueue.prototype.forEach = function(callback) {
-	var keys = Object.keys(this.store).sort();
+/**
+ * Iterates through all the queue elements in priority-then-FIFO order
+ * 
+ * @param {function} callback - Callback function to be used on every element.
+ */
+PriorityQueue.prototype.priorityForEach = function(callback) {
+	for (var priority in this.store){
+		for (var i = 0; i < this.store[priority].length; i++) {
+			callback(this.store[priority][i]);
+		}
+	}
+};
 
-	for (var a = keys.length; a > 0; a--) {
-		for (var b = 0; b < this.store[a].length; b++)
-			callback(this.store[a][b]);
+/**
+ * Tests if a priority's array is empty and deletes the priority if not.
+ * 
+ * @param {Number} priority - Priority to be tested
+ */
+PriorityQueue.prototype.removeEmptyPriority = function(priority) {
+	if (this.store[priority].length === 0){
+		delete this.store[priority];
 	}
-}
+};
 
+/**
+ * Changes the Priority to a different key value.
+ *
+ * @param {*} value - The Item we wish to change
+ * @param {Number} newPriority - Positive integer 
+ */
 PriorityQueue.prototype.changePriority = function(value, newPriority) {
 	var foundItem = false;
 
-	this.store.forEach(function(bucket) {

```

Comment:
This function threw me off. I thought it was trying to use the original `PriorityQueue.prototype.forEach` which didn't make sense since it was being used like the object equivalent of `Array.prototype.forEach` yet `PriorityQueue.prototype.forEach`'s implementation iterated through the entire data structure. I assumed that it would be used for exactly as it's original description and renamed it instead.

Despite the similarities between both this new code and `priorityForEach`, I decided not to leverage the `priorityForEach` because even though I can use the comparison logic in the callback, there isn't any early exit logic that I can use with it. Using this code has slightly lower complexity.